File: Programmers/Lv3/Lv3_예산.py
def solution(budgets, M):
    right = max(budgets)
    left = 0
    
    while left <= right:
        total = 0
        mid = (left+right)>>1
        for i in budgets:
            if i < mid:
                total += i
            else:
                total += mid
        if total > M:
            right = mid-1
        else:
            left = mid + 1
    return right
print(solution([120, 110, 140, 150], 485))      # 127

File: Programmers/Lv3/Lv3_예산.cpp
#include <string>
#include <vector>
using namespace std;

int solution(vector<int> budgets, int M) {
	int answer = -1;
	int sum = 0;
	int temp = 0;
	for (int i = 0; i < budgets.size(); i++) {
		if (answer < budgets[i])
			answer = budgets[i];
		sum += budgets[i];
	}
	temp = sum;
	while (temp > M) {
		temp = sum;
		answer--;
		for (int i = 0; i < budgets.size(); i++) {
			if (answer < budgets[i])
				temp -= budgets[i] - answer;
		}
	}

	return answer;
}
